osbrain/tests/test_nameserver.py

```diff
@@ -77,6 +96,28 @@ def test_nameserver_proxy_shutdown_with_agents():
     ns.shutdown()
 
 
+def test_nameserver_shutdown_with_agents():
+    """
+    Shutdown a name server with agents that wait some time before shutting
+    down.
+    """
+    nameserver, addr = get_nameserver_and_address()
+
+    class NewAgent(Agent):
+        def __init__(self, *args, **kwargs):
+            super(NewAgent, self).__init__(*args, **kwargs)
+
+        @Pyro4.oneway
+        def shutdown(self):
+            time.sleep(0.3)
+            super().shutdown()
+
+    run_agent('Agent0', nsaddr=addr, base=NewAgent)
+    run_agent('Agent1', nsaddr=addr, base=NewAgent)
```

Comment:
Not exactly. I implemented that function since I wanted to directly call `NameServerProcess.shutdown()`. I did some research and couldn't find a way to call it from a NSProxy.

That `get_nameserver_and_address()` function is really a slightly modified `random_nameserver()` (from `nameserver.py` function), main difference being the return value (NameServerProcess itself instead a NSProxy to it).

I have pushed another commit that automatically obtains the address from the NameServerProcess, simplifying the function/test as you suggested :+1: .